File: go/internal/changefile/read_change_files_test.go
package changefile_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"slices"
	"testing"

	"github.com/microsoft/beachball/internal/changefile"
	"github.com/microsoft/beachball/internal/monorepo"
	"github.com/microsoft/beachball/internal/options"
	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func getReadTestOptionsAndPackages(t *testing.T, repo *testutil.Repository, overrides *types.BeachballOptions) (types.BeachballOptions, types.PackageInfos, types.ScopedPackages) {
	t.Helper()
	repoOpts := getDefaultOptions()
	if overrides != nil {
		repoOpts = *overrides
	}
	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, repoOpts)
	packageInfos, err := monorepo.GetPackageInfos(&parsed.Options)
	require.NoError(t, err)
	scopedPackages := monorepo.GetScopedPackages(&parsed.Options, packageInfos)
	return parsed.Options, packageInfos, scopedPackages
}

func getPackageNames(changeSet types.ChangeSet) []string {
	var names []string
	for _, entry := range changeSet {
		names = append(names, entry.Change.PackageName)
	}
	slices.Sort(names)
	return names
}

// TS: "reads change files and returns [them]"
func TestReadChangeFiles_ReadsChangeFiles(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, nil)
	testutil.GenerateChangeFiles(t, []string{"foo", "bar"}, &opts, repo)

	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Len(t, changeSet, 2)
	assert.Equal(t, []string{"bar", "foo"}, getPackageNames(changeSet))
}

// TS: "reads from a custom changeDir"
func TestReadChangeFiles_ReadsFromCustomChangeDir(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	overrides := getDefaultOptions()
	overrides.ChangeDir = "customChanges"
	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, &overrides)
	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)

	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Equal(t, []string{"foo"}, getPackageNames(changeSet))
}

// TS: "reads a grouped change file"
func TestReadChangeFiles_ReadsGroupedChangeFile(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	overrides := getDefaultOptions()
	overrides.GroupChanges = true
	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, &overrides)

	// Write a grouped change file manually
	changePath := changefile.GetChangePath(&opts)
	os.MkdirAll(changePath, 0o755)
	grouped := types.ChangeInfoMultiple{
		Changes: []types.ChangeFileInfo{
			{Type: types.ChangeTypeMinor, Comment: "foo change", PackageName: "foo", Email: "test@test.com", DependentChangeType: types.ChangeTypePatch},
			{Type: types.ChangeTypeMinor, Comment: "bar change", PackageName: "bar", Email: "test@test.com", DependentChangeType: types.ChangeTypePatch},
		},
	}
	data, _ := json.MarshalIndent(grouped, "", "  ")
	os.WriteFile(filepath.Join(changePath, "change-grouped.json"), data, 0o644)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "grouped change file"})

	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Equal(t, []string{"bar", "foo"}, getPackageNames(changeSet))
}

// TS: "excludes invalid change files"
func TestReadChangeFiles_ExcludesInvalidChangeFiles(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	// Make bar private
	barPkgPath := filepath.Join(repo.RootPath(), "packages", "bar", "package.json")
	barData, _ := os.ReadFile(barPkgPath)
	var barPkg map[string]any
	json.Unmarshal(barData, &barPkg)
	barPkg["private"] = true
	barData, _ = json.MarshalIndent(barPkg, "", "  ")
	os.WriteFile(barPkgPath, barData, 0o644)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "make bar private"})

	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, nil)

	// Generate change files: "fake" doesn't exist, "bar" is private, "foo" is valid
	testutil.GenerateChangeFiles(t, []string{"fake", "bar", "foo"}, &opts, repo)

	// Also write a non-change JSON file
	changePath := changefile.GetChangePath(&opts)
	os.WriteFile(filepath.Join(changePath, "not-change.json"), []byte("{}"), 0o644)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "add invalid file"})

	buf := testutil.CaptureLogging(t)
	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Equal(t, []string{"foo"}, getPackageNames(changeSet))

	output := buf.String()
	assert.Contains(t, output, "does not appear to be a change file")
	assert.Contains(t, output, "Change detected for nonexistent package fake; delete this file")
	assert.Contains(t, output, "Change detected for private package bar; delete this file")
}

// TS: "excludes invalid changes from grouped change file in monorepo"
func TestReadChangeFiles_ExcludesInvalidChangesFromGroupedFile(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	// Make bar private
	barPkgPath := filepath.Join(repo.RootPath(), "packages", "bar", "package.json")
	barData, _ := os.ReadFile(barPkgPath)
	var barPkg map[string]any
	json.Unmarshal(barData, &barPkg)
	barPkg["private"] = true
	barData, _ = json.MarshalIndent(barPkg, "", "  ")
	os.WriteFile(barPkgPath, barData, 0o644)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "make bar private"})

	overrides := getDefaultOptions()
	overrides.GroupChanges = true
	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, &overrides)

	// Write a grouped change file with invalid entries
	changePath := changefile.GetChangePath(&opts)
	os.MkdirAll(changePath, 0o755)
	grouped := types.ChangeInfoMultiple{
		Changes: []types.ChangeFileInfo{
			{Type: types.ChangeTypeMinor, Comment: "fake change", PackageName: "fake", Email: "test@test.com", DependentChangeType: types.ChangeTypePatch},
			{Type: types.ChangeTypeMinor, Comment: "bar change", PackageName: "bar", Email: "test@test.com", DependentChangeType: types.ChangeTypePatch},
			{Type: types.ChangeTypeMinor, Comment: "foo change", PackageName: "foo", Email: "test@test.com", DependentChangeType: types.ChangeTypePatch},
		},
	}
	data, _ := json.MarshalIndent(grouped, "", "  ")
	os.WriteFile(filepath.Join(changePath, "change-grouped.json"), data, 0o644)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "grouped change file"})

	buf := testutil.CaptureLogging(t)
	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Equal(t, []string{"foo"}, getPackageNames(changeSet))

	output := buf.String()
	assert.Contains(t, output, "Change detected for nonexistent package fake; remove the entry from this file")
	assert.Contains(t, output, "Change detected for private package bar; remove the entry from this file")
}

// TS: "excludes out of scope change files in monorepo"
func TestReadChangeFiles_ExcludesOutOfScopeChangeFiles(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/file.js")

	overrides := getDefaultOptions()
	overrides.Scope = []string{"packages/foo"}
	opts, infos, scoped := getReadTestOptionsAndPackages(t, repo, &overrides)

	testutil.GenerateChangeFiles(t, []string{"bar", "foo"}, &opts, repo)

	changeSet := changefile.ReadChangeFiles(&opts, infos, scoped)
	assert.Equal(t, []string{"foo"}, getPackageNames(changeSet))
}

File: rust/tests/read_change_files_test.rs
mod common;

use beachball::changefile::read_change_files::{get_change_path, read_change_files};
use beachball::monorepo::package_infos::get_package_infos;
use beachball::monorepo::scoped_packages::get_scoped_packages;
use beachball::options::get_options::get_parsed_options_for_test;
use beachball::types::change_info::{ChangeFileInfo, ChangeInfoMultiple, ChangeType};
use beachball::types::options::{BeachballOptions, CliOptions};
use common::change_files::generate_change_files;
use common::repository_factory::RepositoryFactory;
use common::{
    DEFAULT_BRANCH, DEFAULT_REMOTE_BRANCH, capture_logging, get_log_output, reset_logging,
};
use serde_json::json;
use std::path::Path;

fn make_opts() -> BeachballOptions {
    BeachballOptions {
        branch: DEFAULT_REMOTE_BRANCH.to_string(),
        fetch: false,
        ..Default::default()
    }
}

fn get_options_and_packages(
    repo: &common::repository::Repository,
    overrides: Option<BeachballOptions>,
) -> (
    BeachballOptions,
    beachball::types::package_info::PackageInfos,
    beachball::types::package_info::ScopedPackages,
) {
    let mut opts = overrides.unwrap_or_else(make_opts);
    opts.branch = DEFAULT_REMOTE_BRANCH.to_string();
    opts.fetch = false;
    let parsed = get_parsed_options_for_test(repo.root_path(), CliOptions::default(), opts);
    let infos = get_package_infos(&parsed.options).unwrap();
    let scoped = get_scoped_packages(&parsed.options, &infos);
    (parsed.options, infos, scoped)
}

fn get_package_names(change_set: &beachball::types::change_info::ChangeSet) -> Vec<String> {
    let mut names: Vec<String> = change_set
        .iter()
        .map(|e| e.change.package_name.clone())
        .collect();
    names.sort();
    names
}

// TS: "reads change files and returns [them]"
#[test]
fn reads_change_files() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    let (options, infos, scoped) = get_options_and_packages(&repo, None);
    generate_change_files(&["foo", "bar"], &options, &repo);

    let change_set = read_change_files(&options, &infos, &scoped);
    assert_eq!(change_set.len(), 2);
    assert_eq!(get_package_names(&change_set), vec!["bar", "foo"]);
}

// TS: "reads from a custom changeDir"
#[test]
fn reads_from_custom_change_dir() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    let opts = BeachballOptions {
        change_dir: "customChanges".to_string(),
        ..make_opts()
    };
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts));
    generate_change_files(&["foo"], &options, &repo);

    let change_set = read_change_files(&options, &infos, &scoped);
    assert_eq!(get_package_names(&change_set), vec!["foo"]);
}

// TS: "reads a grouped change file"
#[test]
fn reads_grouped_change_file() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    let opts = BeachballOptions {
        group_changes: true,
        ..make_opts()
    };
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts));

    // Write a grouped change file
    let change_path = get_change_path(&options);
    std::fs::create_dir_all(&change_path).ok();
    let grouped = ChangeInfoMultiple {
        changes: vec![
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "foo change".to_string(),
                package_name: "foo".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "bar change".to_string(),
                package_name: "bar".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
        ],
    };
    let json = serde_json::to_string_pretty(&grouped).unwrap();
    std::fs::write(Path::new(&change_path).join("change-grouped.json"), json).unwrap();
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "grouped change file"]);

    let change_set = read_change_files(&options, &infos, &scoped);
    assert_eq!(get_package_names(&change_set), vec!["bar", "foo"]);
}

// TS: "excludes invalid change files"
#[test]
fn excludes_invalid_change_files() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    // Make bar private
    let bar_pkg = json!({"name": "bar", "version": "1.0.0", "private": true});
    repo.write_file_content(
        "packages/bar/package.json",
        &serde_json::to_string_pretty(&bar_pkg).unwrap(),
    );
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "make bar private"]);

    let (options, infos, scoped) = get_options_and_packages(&repo, None);

    // Generate change files: "fake" doesn't exist, "bar" is private, "foo" is valid
    generate_change_files(&["fake", "bar", "foo"], &options, &repo);

    // Also write a non-change JSON file
    let change_path = get_change_path(&options);
    std::fs::write(Path::new(&change_path).join("not-change.json"), "{}").unwrap();
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "add invalid file"]);

    capture_logging();
    let change_set = read_change_files(&options, &infos, &scoped);
    let output = get_log_output();
    reset_logging();

    assert_eq!(get_package_names(&change_set), vec!["foo"]);
    assert!(output.contains("does not appear to be a change file"));
    assert!(output.contains("Change detected for nonexistent package fake; delete this file"));
    assert!(output.contains("Change detected for private package bar; delete this file"));
}

// TS: "excludes invalid changes from grouped change file in monorepo"
#[test]
fn excludes_invalid_changes_from_grouped_file() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    // Make bar private
    let bar_pkg = json!({"name": "bar", "version": "1.0.0", "private": true});
    repo.write_file_content(
        "packages/bar/package.json",
        &serde_json::to_string_pretty(&bar_pkg).unwrap(),
    );
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "make bar private"]);

    let opts = BeachballOptions {
        group_changes: true,
        ..make_opts()
    };
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts));

    // Write a grouped change file with invalid entries
    let change_path = get_change_path(&options);
    std::fs::create_dir_all(&change_path).ok();
    let grouped = ChangeInfoMultiple {
        changes: vec![
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "fake change".to_string(),
                package_name: "fake".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "bar change".to_string(),
                package_name: "bar".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "foo change".to_string(),
                package_name: "foo".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
        ],
    };
    let json = serde_json::to_string_pretty(&grouped).unwrap();
    std::fs::write(Path::new(&change_path).join("change-grouped.json"), json).unwrap();
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "grouped change file"]);

    capture_logging();
    let change_set = read_change_files(&options, &infos, &scoped);
    let output = get_log_output();
    reset_logging();

    assert_eq!(get_package_names(&change_set), vec!["foo"]);
    assert!(
        output.contains(
            "Change detected for nonexistent package fake; remove the entry from this file"
        )
    );
    assert!(
        output.contains("Change detected for private package bar; remove the entry from this file")
    );
}

// TS: "excludes out of scope change files in monorepo"
#[test]
fn excludes_out_of_scope_change_files() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    let opts = BeachballOptions {
        scope: Some(vec!["packages/foo".to_string()]),
        ..make_opts()
    };
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts));
    generate_change_files(&["bar", "foo"], &options, &repo);

    let change_set = read_change_files(&options, &infos, &scoped);
    assert_eq!(get_package_names(&change_set), vec!["foo"]);
}

// TS: "excludes out of scope changes from grouped change file in monorepo"
#[test]
fn excludes_out_of_scope_changes_from_grouped_file() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);
    repo.commit_change("packages/foo/file.js");

    let opts = BeachballOptions {
        scope: Some(vec!["packages/foo".to_string()]),
        group_changes: true,
        ..make_opts()
    };
    let (options, infos, scoped) = get_options_and_packages(&repo, Some(opts));

    // Write a grouped change file with bar+foo
    let change_path = get_change_path(&options);
    std::fs::create_dir_all(&change_path).ok();
    let grouped = ChangeInfoMultiple {
        changes: vec![
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "bar change".to_string(),
                package_name: "bar".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
            ChangeFileInfo {
                change_type: ChangeType::Minor,
                comment: "foo change".to_string(),
                package_name: "foo".to_string(),
                email: "test@test.com".to_string(),
                dependent_change_type: ChangeType::Patch,
            },
        ],
    };
    let json = serde_json::to_string_pretty(&grouped).unwrap();
    std::fs::write(Path::new(&change_path).join("change-grouped.json"), json).unwrap();
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "grouped change file"]);

    let change_set = read_change_files(&options, &infos, &scoped);
    assert_eq!(get_package_names(&change_set), vec!["foo"]);
}

// Skipped TS tests:
// - "runs transform.changeFiles functions if provided" — transform not implemented
// - "filters change files to only those modified since fromRef" — fromRef not implemented
// - "returns empty set when no change files exist since fromRef" — fromRef not implemented
// - "excludes deleted change files when using fromRef" — fromRef not implemented
